R/callbacks-mixup.R

```diff
@@ -0,0 +1,111 @@
+#' described here: https://forums.fast.ai/t/mixup-data-augmentation/22764).
+#' Namely,
+#' - We work with a single dataloader only, randomly mixing two observations from the same batch.
+#' - We linearly combine losses computed for both targets:
+#'   `loss(output, new_target) = weight * loss(output, target1) + (1-weight) * loss(output, target2)`
+#' - We draw different mixing coefficients for every pair.
+#' - We replace `weight` with `weight = max(weight, 1-weight)` to avoid duplicates.
+#'
+#' @param alpha parameter for the beta distribution used to sample mixing coefficients
+#'
+#' @examples
+#' mixup_callback <- luz_callback_mixup()
+#'
+#' @returns
+#' A `luz_callback`
+#'
+#' @seealso [nn_mixup_loss()], [nnf_mixup()]
+#'
+#' @family luz_callbacks
+#' @export
+luz_callback_mixup <- luz_callback(
+  "mixup_callback",
+  initialize = function(alpha = 0.4) {
+    self$alpha <- alpha
+  },
+
+  on_train_batch_begin = function() {
+
+    batch_len <- ctx$batch$y$size(1)
+    xdim <- length(ctx$batch$x$size())
+    xrep <- rep(1, xdim -1)
+    device <- ctx$batch$y$device
+
+    # draw mixing weights from a beta distribution with identical parameters
+    weight <- rbeta(batch_len, self$alpha, self$alpha) %>% torch::torch_tensor(device = device)
+    weight <- torch::torch_stack(list(weight, 1 - weight), 2)
+    weight <- weight$max(2)[1][[1]]
+
+    # determine which observations to mix
+    shuffle <- torch::torch_randperm(batch_len, dtype = torch::torch_long(), device = device) + 1L
+
+    # (1) linearly combine the inputs according to the mixing weights,
+    # (2) as new target, create a list of:
+    #     - (a) both targets stacked into a single tensor and
+    #     - (b) a tensor holding the mixing weights,
+    # (3) and replace the current batch with this
+    c(mixed_x, stacked_y_with_weights) %<-% nnf_mixup(
+      ctx$batch$x,
+      ctx$batch$y,
+      shuffle,
+      weight$view(list(batch_len, xrep) %>% unlist()))
+
+    ctx$batch$x <- mixed_x
+    ctx$batch$y <- stacked_y_with_weights
+  }
+)
+
+#' Mixup logic
+#'
+#' Logic underlying [luz_callback_mixup()].
+#'
+#' @param x an input batch
+#' @param y a target batch
+#' @param shuffle indices to be used to draw items to complement the "real batch"
+#' @param weight weighting coefficient to be used by `torch_lerp()`
+#'
+#' @examples
+#' batch_x <- torch::torch_randn(c(10, 768))
+#' batch_y <- torch::torch_randn(10)
+#' shuffle <- torch::torch_tensor(10:1)
+#' weight <- torch::torch_tensor(rep(0.9, 10))$view(c(10, 1))
+#' nnf_mixup(batch_x, batch_y, shuffle, weight)
+#'
+#' @returns
+#' A `list` of:
+#' - `x`, the new, mixed-up input batch
+#' - `y`, a `list` of:
+#'   - `ys`, a `list` of:
+#'     - `y1`, the original target `y1`
+#'     - `y2`, the mixed-in target `y2`
+#'   - `weight`, the mixing weights
+#'
+#' @seealso [luz_callback_mixup()]
+#'
+#' @export
+nnf_mixup <- function(x, y, shuffle, weight) {
```

Comment:
I think shuffling should happen entirely inside this function. Would there be any special reason for it to happen outside?

Reply:
Yes, in fact there is ... I wanted `nnf_mixup()` to be deterministic, so it can be tested.

Reply:
@dfalbel I changed my mind on this ... For "practical" testing and visualization, it is enough to have the weights passed-in to the function, but if shuffling remains transparent to the user, that's ok.
As a plus, the function gets easier to use (to the end user).
Just pushed the change.